### ISSUE TITLE
Unify types between packages/app and packages/ui-extensions-dev-server-kit

### ISSUE THREAD

**Issue by byrichardpowell** [Shopify/cli#835]
### TLDR

To prevent extension types betwen `app` and `ui-extensions-server-kit` from falling out of alignment, and to prevent packages from depending on code they shouldn't we will introduce a new private package `ui-extensions-kit`.  This will contain types (and in future maybe utilities) that both `app` and `ui-extensions-server-kit` depend on.

### Context

If we simplify extension development we can say that 3 packages are involved:

* `app`: Loads extensions from disk, validates config, builds source code, watches for changes, maintains some state, exposes a websocket for updates (e.g: hot reloading) and a REST API. This data is consumed by `ui-extensions-dev-console` via `ui-extensions-server-kit`.
* `ui-extensions-dev-console`: Renders a UI to expose the data and features provided by `app` via `ui-extensions-server-kit`.
* `ui-extensions-server-kit`: Provides functionality common to Extension Hosts (e.g: admin, checkout) and the `ui-extensions-dev-console`.

### History

The functionality for `app` used to be provided by a go package, but we moved it's functionality  to `app` and then [deleted the Go package](https://github.com/Shopify/cli/pull/690). This is important because at one point `ui-extensions-server-kit` could not reuse types defined in go. 

### Problem

`ui-extensions-server-kit` defines it's own types.  These are supposed to match the data provided by `app`, but `app` defines it's own types.  These two types have fallen out of alignment.

### Objective

1. Align the types to fix the problem
2. Prevent future issues by making it difficult for these types to fall out of alignment.

#### Option 1: New package (Recommended)

Introduce a new package in CLI called `ui-extensions-kit`.  This package defines the shared types and any utilities that are shared by `ui-extensions-server-kit` and `app`

Pros:

* Solves the problem & satisfies both objectives
* It's a small package, so having other packages depend on it is ok. 

Cons:

* Overhead of another package

#### Option 2: Use app package

Add these types to `app`. `ui-extensions-kit` depends on `app`.

Pros:

* Solves the problem & satisfies both objectives
* Don't have to create a new package

Cons:

* `ui-extensions-kit` and `app` are very different concepts.  One is a UI or server runtime.  The other is a CLI only runtime.  Coupling these two seems like it would be adding cement.
* Potential to make a mistake and accidentally build a whole bunch of CLI code in server kit.

#### Option 3: Use cli-kit package

Add these types to `cli-kit`. `app` and `ui-etensions-server-kit` depends on `app`.

Pros:

* Solves the problem & satisfies both objectives
* Don't have to create a new package

Cons:

* `ui-extensions-kit` and `cli-kit` are very different concepts.  One is a UI or server runtime.  The other is a CLI only runtime.  Coupling these two seems like it would be adding cement.
* `cli-kit` is not supposed to have any domain specific code (e.g: extensions)
* Potential to make a mistake and accidentally build a whole bunch of CLI code in server kit.

**Comment by byrichardpowell** [comment]
See: https://github.com/Shopify/cli/pull/833
